packages/backend-cli/src/directLine.ts

```diff
@@ -0,0 +1,772 @@
+
+interface ActivityGroup {
+  activities: Activity[],
+  watermark: string
+}
+
+// These types are specific to this client library, not to Direct Line 3.0
+
+export enum ConnectionStatus {
+  Uninitialized,              // the status when the DirectLine object is first created/constructed
+  Connecting,                 // currently trying to connect to the conversation
+  Online,                     // successfully connected to the converstaion. Connection is healthy so far as we know.
+  ExpiredToken,               // last operation errored out with an expired token. Possibly waiting for someone to supply a new one.
+  FailedToConnect,            // the initial attempt to connect to the conversation failed. No recovery possible.
+  Ended                       // the bot ended the conversation
+}
+
+export type Fetch = (input?: string | Request, init?: RequestInit) => Promise<Response>
+export interface FormDataAttachment {
+  contentType?: string;
+  data: any;
+  filename?: string;
+  name: string;
+}
+
+export interface DirectLineOptions {
+  secret?: string;
+  token?: string;
+  conversationId?: string;
+  watermark?: string;
+  domain?: string;
+  webSocket?: boolean;
+  pollingInterval?: number;
+  streamUrl?: string;
+  fetch: Fetch;
+  createFormData: (attachments: FormDataAttachment[]) => any;
+}
+
+const lifetimeRefreshToken = 30 * 60 * 1000;
+const intervalRefreshToken = lifetimeRefreshToken / 2;
+const timeout = 20 * 1000;
+const retries = (lifetimeRefreshToken - intervalRefreshToken) / timeout;
+
+const errorExpiredToken = new Error("expired token");
+const errorConversationEnded = new Error("conversation ended");
+const errorFailedToConnect = new Error("failed to connect");
+
+const konsole = {
+  log: (message?: any, ... optionalParams: any[]) => {
+    if (typeof(window) !== 'undefined' && window["botchatDebug"] && message)
+      console.log(message, ... optionalParams);
+  }
+}
+
+export interface IBotConnection {
+  connectionStatus$: BehaviorSubject<ConnectionStatus>,
+  activity$: Observable<Activity>,
+  end(): void,
+  referenceGrammarId?: string,
+  postActivity(activity: Activity): Observable<string>,
+  getSessionId? : () => Observable<string>
+}
+
+function rejectAfter(timeout: number) {
+  return new Promise((_, reject) => {
+    setTimeout(() => {
+      reject(new Error('timeout'));
+    }, timeout);
+  });
+}
+
+function fetchAsObservable(fetch: Fetch, options: any): Observable<{ response: ArrayBuffer | any }> {
+  const {
+    responseType,
+    timeout,
+    url,
+    withCredentials,
+    ...fetchOptions
+  } = options;
+
+  if (withCredentials) {
+    fetchOptions.credentials = 'include';
+  }
+
+  return Observable.defer(
+    () => Promise.race([
+      fetch(url, fetchOptions),
+      timeout ? rejectAfter(timeout) : null
+    ]).then(async (res: Response) => {
+      if (!res.ok) {
+        throw new Error(res.statusText || (`server returned ${ res.status }`));
```

Comment:
this throw, and the rejectAfter, will reject the promise here? What does the observer get?